### PR TITLE
std::vector based explodelist

### PR DESCRIPTION
**Pull request by bmorel**
* new explodelist that can emit std::vector<std::string>>
* use it when it does not imply API change

### CODE REVIEW

File: src/engine/rendertext.cpp [hunk 1]
```diff
@@ -712,26 +714,25 @@ float key_widthf(const char *str)
...
+    std::vector<std::string> list;
     explodelist(keyn, list);
     float width = 0, scale = curfont->maxh*curfont->scale/float(curfont->defaulth)*curtextscale*textkeyimagescale;
     loopv(list)
     {
         if(i && textkeyseps) width += text_widthf("|");
         if(textkeyimages)
         {
-            textkey *t = findtextkey(list[i]);
+            textkey *t = findtextkey(list[i].data());
```
Comment (MoonPadUSer) [inline src/engine/rendertext.cpp:725]: same here, if you are editing the line anyways, please change it to `textkey* t`

File: src/engine/rendertext.cpp [hunk 2]
```diff
@@ -740,14 +741,14 @@ static int draw_key(Texture *&tex, const char *str, float sx, float sy)
...
-    vector<char *> list;
+    std::vector<std::string> list;
     explodelist(keyn, list);
     float width = 0, sh = curfont->maxh*curfont->scale/float(curfont->defaulth)*curtextscale, h = sh*textkeyimagescale;
     loopv(list)
     {
         if(textkeyimages)
         {
-            textkey *t = findtextkey(list[i]);
+            textkey *t = findtextkey(list[i].data());
```
Comment (MoonPadUSer) [inline src/engine/rendertext.cpp:751]: same as above

File: src/engine/skelmodel.h [hunk 3]
```diff
@@ -1810,17 +1810,20 @@ template<class MDL> struct skelcommands : modelcommands<MDL, struct MDL::skelmes
...
             int bone = p->meshes ? ((meshgroup *)p->meshes)->skel->findbone(bonestr[0]=='!' ? bonestr+1 : bonestr) : -1;
-            if(bone<0) { conoutf("\frcould not find bone %s for anim part mask [%s]", bonestr, maskstr); bonestrs.deletearrays(); return; }
+            if(bone<0)
+            {
+                conoutf("\frcould not find bone %s for anim part mask [%s]", bonestr, maskstr);
+                return;
+            }
             bonemask.add(bone | (bonestr[0]=='!' ? BONEMASK_NOT : 0));
         }
-        bonestrs.deletearrays();
```
Comment (MoonPadUSer) [inline src/engine/skelmodel.h:1823]: why did you remove this here instead of replacing it with `bonestr.clear()` like you did in other files too?

File: src/engine/command.cpp [hunk 4]
```diff
@@ -2731,6 +2734,13 @@ void explodelist(const char *s, vector<char *> &elems, int limit)
         elems.add(newstring(start, end-start));
 }
 
+void explodelist(const char *s, std::vector<std::string> &elems, int limit)
```
[the remark this answers is not in the document]
Comment (bmorel) [reply src/engine/command.cpp:2737]: where was it said?

[the remark this answers is not in the document]
Comment (MoonPadUSer) [reply src/engine/command.cpp:2737]: I think it was said on IRC when we talked about wether or not it's written` int *a` or `int* a`, you and TA agreed that the correct form is `int* a` so I assumed that's gonna be what we'll use

File: src/game/server.cpp [hunk 5]
```diff
@@ -3373,32 +3379,33 @@ namespace server
...
             {
-                vector<char *> prev;
+                std::vector<std::string> prev;
                 explodelist(sv_previousmaps, prev);
-                loopvrev(prev) if(!strcmp(prev[i], smapname))
+                for( ssize_t i = prev.size() - 1; i >= 0; --i )
                 {
-                    delete[] prev[i];
-                    prev.remove(i);
+                    if(!strcmp(prev[i].data(), smapname))
```
Comment (TheAssassin) [inline src/game/server.cpp:3390]: No need for C string stuff.

File: src/game/server.cpp [hunk 6]
```diff
@@ -3373,32 +3379,33 @@ namespace server
...
-            buf.put(smapname, strlen(smapname));
+            std::vector<char> buf;
+            buf.insert( buf.end(), smapname, smapname + strlen(smapname));
             if(*sv_previousmaps && G(maphistory))
             {
-                vector<char *> prev;
+                std::vector<std::string> prev;
                 explodelist(sv_previousmaps, prev);
-                loopvrev(prev) if(!strcmp(prev[i], smapname))
+                for( ssize_t i = prev.size() - 1; i >= 0; --i )
```
Comment (TheAssassin) [inline src/game/server.cpp:3388]: Reverse iterators ftw!

File: src/shared/command.h [hunk 7]
```diff
@@ -349,6 +351,7 @@ static inline const char *escapeid(ident &id) { return escapeid(id.name); }
 extern bool validateblock(const char *s);
 extern char *parsetext(const char *&p);
 extern void explodelist(const char *s, vector<char *> &elems, int limit = -1);
```
Comment (TheAssassin) [inline src/shared/command.h:353]: That prototype should no longer be needed, right?

File: src/game/weapons.cpp [hunk 8]
```diff
@@ -21,22 +23,21 @@ namespace weapons
...
-        vector<char *> list;
+        std::vector<std::string> list;
         explodelist(str, list);
         weaplist.shrink(0);
         loopv(list)
         {
             int weap = -1;
-            if(isnumeric(list[i][0])) weap = atoi(list[i]);
-            else loopj(W_ALL) if(!strcasecmp(weaptype[j].name, list[i]))
+            if(isnumeric(list[i][0])) weap = atoi(list[i].data());
```
Comment (TheAssassin) [inline src/game/weapons.cpp:32]: Would appreciate if you could clean up this "stuff everything into one line" stuff.

File: src/engine/command.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (bmorel) [reply src/engine/command.cpp:2737]: Why do you think I know what's the intended purpose? I identified this code to be the source of memory leaks, and rewrote it so that it would prevent some of the leaks (by using STL), that's all.

File: src/game/client.cpp [hunk 9]
```diff
@@ -398,11 +400,10 @@ namespace client
...
     void updateserversort()
     {
-        vector<char *> styles;
+        std::vector<std::string> styles;
         explodelist(serversort, styles, SINFO_MAX);
-        serversortstyles.setsize(0);
-        loopv(styles) serversortstyles.add(parseint(styles[i]));
-        styles.deletearrays();
+        serversortstyles.clear();
+        for( size_t i = 0; i < styles.size(); ++i ) serversortstyles.emplace_back(parseint(styles[i].data()));
```
[the remark this answers is not in the document]
Comment (bmorel) [reply src/game/client.cpp:406]: Alright, I'll undo every change that is not directly linked to explodelist.

File: src/shared/command.h (hunk 7 above)
Comment (TheAssassin) [reply src/shared/command.h:353]: Could we flag it with a notice that says "deprecated, please use new `std::vector` version"? I think doxygen has a syntax for that.

File: src/engine/command.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (TheAssassin) [reply src/engine/command.cpp:2737]: Maybe we should document *that* then in a docstring. It'd be helpful to document the bits we do not (yet) understand, too.

File: src/game/client.cpp [hunk 10]
```diff
@@ -496,15 +496,14 @@ namespace client
...
+            std::vector<std::string> chunk;
             explodelist(list, chunk);
             loopv(chunk)
             {
-                if(!chunk[i] || !*chunk[i] || !isnumeric(*chunk[i])) continue;
-                int v = parseint(chunk[i]);
-                items.add(v ? 1 : 0);
+                if(chunk[i].empty() || !isnumeric(chunk[i][0])) continue;
+                int v = parseint(chunk[i].c_str());
+                items.emplace_back(v >= W_OFFSET && v < W_ITEM ? v : 0);
```
Comment (TheAssassin) [inline src/game/client.cpp:505]: This comparison has been changed significantly. Will have to check the values of `W_OFFSET` and `W_ITEM` before approving this PR.

File: src/game/server.cpp [hunk 11]
```diff
@@ -1058,29 +1060,27 @@ namespace server
...
                     }
-                    prev.deletearrays();
+                    prev.clear();
                 }
                 if(!maps.empty())
                 {
-                    int r = rnd(maps.length());
-                    copystring(defaultmap, maps[r]);
+                    int r = rnd(maps.size());
+                    defaultmap = maps[r];
```
Comment (TheAssassin) [inline src/game/server.cpp:1080]: Given the parameter order in libc, one could expected `defaultmap` to be the target. But I'll have to double-check.

File: src/game/server.cpp [hunk 12]
```diff
@@ -3373,33 +3373,27 @@ namespace server
...
-                    int last = prev.length()-1;
-                    delete[] prev[last];
-                    prev.remove(last);
+                    prev.pop_back();
                 }
                 loopv(prev)
                 {
-                    buf.add(' ');
-                    buf.put(prev[i], strlen(prev[i]));
+                    buf += ' ';
```
Comment (TheAssassin) [inline src/game/server.cpp:3391]: A stringstream might've been easier/more efficient.

File: src/game/server.cpp (hunk 11 above)
Comment (TheAssassin) [reply src/game/server.cpp:1080]: Yeah, looks like it.
